Path-Check/covid-safe-paths
# Feature: Logging Location via QR Codes

#### Description:

This feature would allow the integration of QR codes with SafePaths. The idea is that this feature could supplement automatic GPS logging in situations where GPS data might be unavailable or imprecise. Locations which support this feature would use a QR code generator (such as [this demo](https://safepaths-qr-6b479.web.app/) I made) and then display this QR code publicly. People would be able to scan these with their phone. Currently, this supports both deep linking (scanning with an external QR code reader will directly open in the app) and scanning from within the app itself.

Of note, this does not currently support universal linking (a feature which would use a web url to handle the link, which would direct users to the app if they already have it, or to the app store otherwise). I held off on universal linking because there could be privacy concerns there. (I.e. because the QR code has to contain location coordinates as url parameters, technically these would be getting sent to a server somewhere where they could potentially be associated with IP addresses.) Deep linking directly to the app, as I'm doing currently, doesn't have this issue.

- Source code for the QR code generator: https://github.com/tyleryasaka/safe-paths-qr-gen
- QR code generator demo: https://safepaths-qr-6b479.web.app/

#### Linked issues:

N/A

#### Screenshots:

![Screenshot_20200509-150014](https://user-images.githubusercontent.com/6504519/81485920-eb655580-9205-11ea-9dd7-d37f79472324.png)

![Screenshot_20200507-141018](https://user-images.githubusercontent.com/6504519/81345088-b40f7100-906c-11ea-88fe-12d58ce15fac.png)

#### How to test:

##### 1. Testing externally scanned QR codes (deep linking)

- For Android, run this command once you have the app running: `adb shell am start -W -a android.intent.action.VIEW -d "safepaths://qr/34.56/-123.45" org.pathcheck.covidsafepaths`
- For iOS, run this command once you have the app running: `xcrun simctl openurl booted safepaths://qr/34.56/-123.45`
- You should see a screen saying that the scan was successful.
- Tap the "home" button on this screen to return to the main screen. Now tap the settings icon and export your location history. Examine this file. You should see an entry with coordinates corresponding to the latitude 34.56 and longitude -123.45 (from the dummy url above).

##### 2. Testing QR code scanning within the app

- First, go to https://safepaths-qr-6b479.web.app/ and follow the steps to generate a QR code.
- Open the app and tap the QR code icon next to the settings icon in the top-right-hand corner of the screen. Scan the QR code generated above (or alternatively, to see the error screen, scan any other QR code). You should see a screen saying that the scan was successful (or that it failed, if you scanned an invalid code).
- Tap the "home" button on this screen to return to the main screen. Now tap the settings icon and export your location history. Examine this file. You should see an entry with coordinates corresponding to the location you used when generating the QR code.
